container_images/concourse-metrics/pkg/requests/requests_test.go

```diff
@@ -0,0 +1,130 @@
+	assertEqual(t, result.TimeSeries[0].Resource.Labels["location"], zone)
+	assertEqual(t, result.TimeSeries[0].Resource.Labels["namespace"], pipeline)
+	assertEqual(t, result.TimeSeries[0].Resource.Labels["job"], job)
+	assertEqual(t, result.TimeSeries[0].Resource.Labels["task_id"], task)
+
+	endTimestamp := result.TimeSeries[0].Points[0].Interval.EndTime.Seconds
+	nowSeconds := time.Now().Unix()
+	if endTimestamp < nowSeconds {
+		t.Errorf("Expected default EndTimestamp to be greater than or equal to now. Result: %v Expected: %v", endTimestamp, nowSeconds)
+	}
+}
+
+func TestBuildJobResultRequestInputValidation(t *testing.T) {
+	// Instantiated since you can't make pointers to int literals (since they are not addressable).
+	validEndTimestamp := int64(1620363568000)
+
+	tests := []struct {
+		args    JobResultArgs
+		message string
+	}{
+		{JobResultArgs{EndTimestamp: &validEndTimestamp, MetricPath: metricPath, StartTimestamp: startTimestamp, Job: job, Pipeline: pipeline, ProjectID: projectID, ResultState: resultState, Task: task, Zone: " "}, "empty zone"},
+		{JobResultArgs{EndTimestamp: &validEndTimestamp, MetricPath: metricPath, StartTimestamp: startTimestamp, Job: job, Pipeline: pipeline, ProjectID: projectID, ResultState: resultState, Task: " ", Zone: zone}, "empty task"},
+		{JobResultArgs{EndTimestamp: &validEndTimestamp, MetricPath: metricPath, StartTimestamp: startTimestamp, Job: job, Pipeline: pipeline, ProjectID: projectID, ResultState: "other", Task: task, Zone: zone}, "invalid state"},
+		{JobResultArgs{EndTimestamp: &validEndTimestamp, MetricPath: metricPath, StartTimestamp: startTimestamp, Job: job, Pipeline: pipeline, ProjectID: " ", ResultState: resultState, Task: task, Zone: zone}, "empty project"},
+		{JobResultArgs{EndTimestamp: &validEndTimestamp, MetricPath: metricPath, StartTimestamp: startTimestamp, Job: job, Pipeline: " ", ProjectID: projectID, ResultState: resultState, Task: task, Zone: zone}, "empty pipeline"},
+		{JobResultArgs{EndTimestamp: &validEndTimestamp, MetricPath: metricPath, StartTimestamp: startTimestamp, Job: " ", Pipeline: pipeline, ProjectID: projectID, ResultState: resultState, Task: task, Zone: zone}, "empty job"},
+		{JobResultArgs{EndTimestamp: &validEndTimestamp, MetricPath: metricPath, StartTimestamp: validEndTimestamp + 1000, Job: job, Pipeline: pipeline, ProjectID: projectID, ResultState: resultState, Task: task, Zone: zone}, "end-timestamp cannot occur before start-timestamp"},
+	}
+	for _, test := range tests {
+		if _, err := BuildJobResultRequest(test.args); err == nil || !strings.Contains(err.Error(), test.message) {
+			t.Errorf("BuildJobResultRequest(%+v) = (_, %v); want error to contain %s", test.args, err, test.message)
+		}
+	}
+}
+
+func assertEqual(t *testing.T, x interface{}, y interface{}) {
+	if x == y {
+		return
+	}
+
+	t.Errorf("Received %v (type %v), expected %v (type %v)", x, reflect.TypeOf(x), y, reflect.TypeOf(y))
```

Comment:
just for future ref, i think you can tell you've gone down a wrong path when you have to do things like this. you've really written this test in a style from some other language, idk any golang testing that looks like this.

Reply:
I think this may be more common in go than you think (a popular framework makes use of it - the one suggested by Zoran): https://github.com/stretchr/testify/blob/master/assert/assertions.go

Reply:
yeah zoran and i disagree here haha, i never use stretchr/testify